### PR TITLE
Misc cleanups and Speedups around sort queries

### PR DESCRIPTION
**Pull request by original-brownbear**
From going through the query parsing logic: 
We can get a few small allocation savings from saving intermediary lists in the sort builder, some minor speedups and code-shorting by using records in a few more spots and a couple more small improvements to save some code + cycles.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/sort/SortFieldAndFormat.java [hunk 1]
```diff
@@ -13,10 +13,7 @@
 
 import java.util.Objects;
 
-public final class SortFieldAndFormat {
-
-    public final SortField field;
-    public final DocValueFormat format;
+public record SortFieldAndFormat(SortField field, DocValueFormat format) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/sort/SortFieldAndFormat.java:16]: shall we extract making this a record from this PR? Otherwise it's hard to follow what else you are changing in the same PR.

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/search/sort/SortFieldAndFormat.java:16]: Right fair point let me do that real quick :)

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/search/sort/SortFieldAndFormat.java:16]: There we go :) https://github.com/elastic/elasticsearch/pull/113398

File: server/src/main/java/org/elasticsearch/search/sort/SortAndFormats.java [hunk 2]
```diff
@@ -11,17 +11,12 @@
 import org.apache.lucene.search.Sort;
 import org.elasticsearch.search.DocValueFormat;
 
-public final class SortAndFormats {
+public record SortAndFormats(Sort sort, DocValueFormat[] formats) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/sort/SortAndFormats.java:14]: this and the other change are still causing noise in this PR. Could you make these a record in a specific PR where nothing needs review as it's only a mechanical change?
